cpp/src/arrow/compute/kernels/scalar_if_else.cc

```diff
@@ -1031,7 +1183,16 @@ void AddPrimitiveIfElseKernels(const std::shared_ptr<ScalarFunction>& scalar_fun
         internal::GenerateTypeAgnosticPrimitive<ResolveIfElseExec,
                                                 /*AllocateMem=*/std::false_type>(*type);
     // cond array needs to be boolean always
-    ScalarKernel kernel({boolean(), type, type}, type, exec);
+    std::shared_ptr<KernelSignature> sig;
+    if (type->id() == Type::TIMESTAMP) {
+      auto unit = checked_cast<const TimestampType&>(*type).unit();
+      sig = KernelSignature::Make(
+          {boolean(), match::TimestampTypeUnit(unit), match::TimestampTypeUnit(unit)},
```

Comment:
Ah, the match here is because otherwise, the kernel for `timestamp[s]` doesn't match `timestamp[s, tz=...]` since the check is exact.

Reply:
Mismatched timezones get caught by CheckIdenticalTypes. Mismatched units get either implicitly casted, or else you get NotImplemented if you have say `timestamp[s]` and `timestamp[ms, tz=America/Phoenix]` since no kernel is registsered for this combo. There are some tests for this.